vmdb/spec/models/authenticator/amazon_spec.rb

```diff
@@ -0,0 +1,461 @@
+require "spec_helper"
+require "aws-sdk"
+
+describe Authenticator::Amazon do
+  subject { Authenticator::Amazon.new(config) }
+  let!(:alice) { FactoryGirl.create(:user, :userid => 'alice') }
+  let(:config) do
+    {
+      :amazon_role   => false,
+      :amazon_key    => 'awskey',
+      :amazon_secret => 'awssecret',
+    }
+  end
+
+  class FakeAmazon
+    def initialize(user_data, *connect_args)
+      @user_data = user_data
+      _connect *connect_args
+    end
+
+    def _connect(username, password, service = nil)
+      if @user_data[username] && @user_data[username][:password] == password
+        @user = username
+      else
+        raise AWS::EC2::Errors::AuthFailure
+      end
+    end
+
+    def regions
+      []
+    end
+
+    def users
+      @user_data.reject { |k, _v| k =~ /aws/ }.map { |k, v| FakeUser.new(k, v) }
+    end
+
+    def client
+      iam_username =
+        case @user_data[@user][:iam]
+        when false
+          nil
+        when :denied
+          raise AWS::IAM::Errors::AccessDenied
+        else
+          @user
+        end
+
+      Object.new.tap do |obj|
+        class << obj
+          attr_accessor :username
+          def get_user
+            {:user => {:user_name => username}}
+          end
+        end
+        obj.username = iam_username
+      end
+    end
+
+    class FakeUser < Struct.new(:username, :data)
+      def name
+        data[:name]
+      end
+
+      def access_keys
+        data[:access_keys].map { |key| FakeKey.new(key) }
+      end
+
+      def groups
+        data[:groups].map { |group| FakeGroup.new(group) }
+      end
+    end
+
+    class FakeKey < Struct.new(:id)
+    end
+
+    class FakeGroup < Struct.new(:name)
+    end
+  end
```

Comment:
I don't think it should be needed outside this file; anything higher level, I'd expect to use a dummy authenticator, or at least mock out the authenticator's body. This should be the only time the authenticator itself needs to believe it's talking to Actual Amazon.

We can always pull it out if/when we need it.